lib/util.js

```diff
@@ -1057,3 +1057,80 @@ exports._exceptionWithHostPort = function(err,
 // process.versions needs a custom function as some values are lazy-evaluated.
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+function onFulfilled({ret, options, cb}) {
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  if ('toSplat' in options && 'project' in options)
+    throw new TypeError('callbackify options can specify either ' +
+                        '`toSplat` or `project`, but not both');
+  const toSplat = options.toSplat;
+  const project = options.project;
+
+
+  if (toSplat) {
+    cb(null, ...ret);
+  } else if (project) {
+    const allThere = project.all((k) => k in ret);
+    if (!allThere) {
+      const notAll = new TypeError('Not all of the projected properties' +
+        ' are present in the returned value:' +
+        inspect(ret));
+      return cb(notAll);
+    } else {
+      const args = project.reduce((s, k) => s.concat(ret[k]), []);
+      cb(null, ...args);
+    }
+  } else {
+    cb(null, ret);
+  }
+}
+
+function onRejected({rej, cb}) {
+  // !rej guard inspired by BlueBird https://goo.gl/t5IS6M
+  if (!rej) {
+    const newRej = new Error(rej + '');
+    newRej.cause = rej;
+    rej = newRej;
+  }
+  cb(rej);
+}
+
+/**
+ * @typedef {Object} callbackifyOptions
+ * @property {function} transformer
+ * @property {boolean} toSplat
+ * @property {Array.<String>} project
+ */
+
+/**
+ * @param {function(): Promise} orig
+ * @param {?callbackifyOptions} options
+ */
+function callbackify(orig, options = {}) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to callbackify() ' +
+                        'must be a function');
+  }
+
+  const fn = function(...args) {
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
+    }
+    var p = orig.call(this, ...args);
+    if (typeof options.transformer === 'function') {
+      p = options.transformer(p);
+    }
+    p = p.then((ret) => process.nextTick(onFulfilled, {ret, options, cb}),
+               (rej) => process.nextTick(onRejected, {rej, cb}));
```

Comment:
Is there any reason for creating objects here? Just passing them as 2 or 3 extra arguments to `nextTick` should be fine